File: src/main/java/ca/alexgirard/harpoonIJ/HarpoonState.java
package ca.alexgirard.harpoonIJ;

import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.LocalFileSystem;
import com.intellij.openapi.vfs.VirtualFile;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class HarpoonState {
    private static final Map<String, List<VirtualFile>> FilesMap = new HashMap<>();
    private static final String ListPersistenceKey = "HarpoonJumpList";

    public static void SetItem(VirtualFile file, int index, Project project) {
        FillLists(project);

        var Files = FilesMap.getOrDefault(project.getName(), new ArrayList<>());
        if (index < 0) {
            Files.add(file);
            return;
        }

        if (index >= Files.size()) {
            for (int i = Files.size(); index >= i; i++) {
                Files.add(null);
            }
        }
        Files.set(index, file);

        var propsComp = PropertiesComponent.getInstance(project);
        var stringLists = Files.stream().filter(virtualFile -> virtualFile != null && virtualFile.isValid()).map(VirtualFile::getPath).toList();
        propsComp.setList(ListPersistenceKey, stringLists);
        FilesMap.put(project.getName(), Files);
    }

    public static VirtualFile GetItem(int index, Project project) {
        FillLists(project);
        var Files = FilesMap.getOrDefault(project.getName(), new ArrayList<>());
        if (index < Files.size()) {
            return Files.get(index);
        }
        return null;

    }

    private static void FillLists(Project project) {
        if (FilesMap.containsKey(project.getName()))//&& FileStringsMap.containsKey(project.getName()))
            return;
        var files = FilesMap.getOrDefault(project.getName(), new ArrayList<>());
        var propsComp = PropertiesComponent.getInstance(project);
        var list = propsComp.getList(ListPersistenceKey);
        list = (list == null) ? new ArrayList<>() : list;
        var LFS = LocalFileSystem.getInstance();
        files = list.stream().map(LFS::findFileByPath).collect(Collectors.toList());
        FilesMap.put(project.getName(), files);
    }

    public static List<VirtualFile> GetFiles(Project project) {
        FillLists(project);
        return FilesMap.getOrDefault(project.getName(), new ArrayList<>());

    }

    private static void ClearLists(Project project) {
        FillLists(project);
        if (FilesMap.containsKey(project.getName()))
            FilesMap.get(project.getName()).clear();
    }

    public static void SetFiles(List<String> list, Project project) {
        ClearLists(project);
        var LFS = LocalFileSystem.getInstance();
        var index = 0;

        for (String s : list) {
            if (s == null || s.isBlank())
                continue;
            var vf = LFS.findFileByPath(s);
            if (vf != null) {
                SetItem(vf, index, project);
                index += 1;
            }
        }
    }
}

File: src/main/kotlin/ca/alexgirard/harpoonIJ/HarpoonState.kt
package ca.alexgirard.harpoonIJ

import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.vfs.VirtualFile
import java.util.stream.Collectors

object HarpoonState {
    private val FilesMap: MutableMap<String, MutableList<VirtualFile?>> = HashMap()
    private val FileStringsMap: MutableMap<String, MutableList<String?>> = HashMap()
    private const val ListPersistenceKey = "HarpoonJumpList"
    fun setItem(file: VirtualFile?, index: Int, project: Project) {
        FillLists(project)
        val Files = FilesMap.getOrDefault(project.name, ArrayList())
        val FileStrings = FileStringsMap.getOrDefault(project.name, ArrayList())
        if (index < 0) {
            Files.add(file)
            FileStrings.add(file!!.path)
            return
        }
        if (index >= Files.size) {
            var i = FileStrings.size
            while (index >= i) {
                Files.add(null)
                FileStrings.add("")
                i++
            }
        }
        Files[index] = file
        val filePath = file?.path
        FileStrings[index] = filePath
        val propsComp = PropertiesComponent.getInstance(project)
        propsComp.setList(ListPersistenceKey, FileStrings)
        FilesMap[project.name] = Files
        FileStringsMap[project.name] = FileStrings
    }

    fun GetItem(index: Int, project: Project): VirtualFile? {
        FillLists(project)
        val Files: List<VirtualFile?> = FilesMap.getOrDefault(project.name, ArrayList())
        return if (index < Files.size) {
            Files[index]
        } else null
    }

    private fun FillLists(project: Project) {
        if (FilesMap.containsKey(project.name) && FileStringsMap.containsKey(project.name)) return
        var Files: MutableList<VirtualFile?>
        var FileStrings: MutableList<String?>? = FileStringsMap.getOrDefault(project.name, ArrayList())
        val propsComp = PropertiesComponent.getInstance(project)
        var list = propsComp.getList(ListPersistenceKey)
        list = list ?: ArrayList()
        val LFS = LocalFileSystem.getInstance()
        Files = list.stream().map { path: String? -> LFS.findFileByPath(path!!) }.collect(Collectors.toList())
        FileStrings = list
        FilesMap[project.name] = Files
        FileStringsMap[project.name] = FileStrings
    }

    @JvmStatic
    fun files(project: Project): List<VirtualFile?> {
        FillLists(project)
        return FilesMap.getOrDefault(project.name, ArrayList())
    }

    private fun clearLists(project: Project) {
        FillLists(project)
        if (FilesMap.containsKey(project.name)) FilesMap[project.name]!!.clear()
        if (FileStringsMap.containsKey(project.name)) FileStringsMap[project.name]!!.clear()
    }

    fun setFiles(list: List<String?>, project: Project) {
        clearLists(project)
        val LFS = LocalFileSystem.getInstance()
        for (i in list.indices) {
            val s = list[i]
            if (s.isNullOrBlank()) setItem(null, i, project) else setItem(LFS.findFileByPath(s), i, project)
        }
    }
}
